packages/tokens/non-fungible/src/extensions/royalties/storage.rs

```diff
@@ -0,0 +1,147 @@
+use soroban_sdk::{contracttype, panic_with_error, Address, Env};
+use stellar_constants::{OWNER_EXTEND_AMOUNT, OWNER_TTL_THRESHOLD};
+
+use crate::{non_fungible::NonFungibleTokenError, Base};
+
+/// Storage container for royalty information
+#[contracttype]
+pub struct RoyaltyInfo {
+    pub receiver: Address,
+    pub basis_points: u32,
+}
+
+/// Storage keys for royalty data
+#[contracttype]
+pub enum RoyaltyStorageKey {
+    DefaultRoyalty,
+    TokenRoyalty(u32),
+}
+
+impl Base {
+    /// Sets the global default royalty information for the entire collection.
+    /// This will be used for all tokens that don't have specific royalty
+    /// information.
+    ///
+    /// # Arguments
+    ///
+    /// * `e` - Access to the Soroban environment.
+    /// * `receiver` - The address that should receive royalty payments.
+    /// * `basis_points` - The royalty percentage in basis points (100 = 1%,
+    ///   10000 = 100%).
+    ///
+    /// # Errors
+    ///
+    /// * [`NonFungibleTokenError::RoyaltyTooHigh`] - If the royalty percentage
+    ///   exceeds the maximum allowed value.
+    ///
+    /// # Events
+    ///
+    /// * topics - `["royalty", receiver: Address]`
+    /// * data - `[basis_points: u32]`
+    ///
+    /// # Notes
+    ///
+    /// **IMPORTANT**: This function lacks authorization controls. Most likely,
+    /// you want to invoke it from a constructor or from another function
+    /// with admin-only authorization.
+    pub fn set_default_royalty(e: &Env, receiver: &Address, basis_points: u32) {
+        // Store the default royalty information
+        let key = RoyaltyStorageKey::DefaultRoyalty;
+        let royalty_info = RoyaltyInfo { receiver: receiver.clone(), basis_points };
+        e.storage().instance().set(&key, &royalty_info);
+    }
+
+    /// Sets the royalty information for a specific token.
+    /// This must be called during minting, as royalties are immutable after
+    /// minting.
+    ///
+    /// # Arguments
+    ///
+    /// * `e` - Access to the Soroban environment.
+    /// * `token_id` - The identifier of the token.
+    /// * `receiver` - The address that should receive royalty payments.
+    /// * `basis_points` - The royalty percentage in basis points (100 = 1%,
+    ///   10000 = 100%).
+    ///
+    /// # Errors
+    ///
+    /// * [`NonFungibleTokenError::NonExistentToken`] - If the token does not
+    ///   exist.
+    /// * [`NonFungibleTokenError::RoyaltyTooHigh`] - If the royalty percentage
+    ///   exceeds the maximum allowed value.
+    /// * [`NonFungibleTokenError::RoyaltyAlreadySet`] - If attempting to set
+    ///   royalties for a token that already has royalty information.
+    ///
+    /// # Events
+    ///
+    /// * topics - `["tokroyal", token_id: u32]`
+    /// * data - `[receiver: Address, basis_points: u32]`
```

Comment:
- this event is not emitted
- I'd suggest renaming the 1st topic to `token_royalty`